zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -146,6 +182,44 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
             &mut reversion_info,
         );
 
+        // Verify transfer
+        let sender_balance_prev = transfer_with_gas_fee.sender.balance_prev();
+
+        let add_tx_value_and_mul_gas_fee_by_gas = cb.query_word();
+
+        // tx_value.int_value + gas_fee.int_value
+        let add_tx_value_and_mul_gas_fee_by_gas_gadget = AddWordsGadget::construct(
+            cb,
+            [tx_value.clone(), mul_gas_fee_by_gas.product().clone()],
+            add_tx_value_and_mul_gas_fee_by_gas.clone(),
+        );
+
+        // balance_not_enough = sender_balance_prev.int_value < tx_value.int_value +
+        // gas_fee.int_value
+        let balance_not_enough = LtWordGadget::construct(
+            cb,
+            &sender_balance_prev,
+            &add_tx_value_and_mul_gas_fee_by_gas,
+        );
+
+        // invalid_tx = 1 - (1 - balance_not_enough) * (is_nonce_valid)
+        // prover should not give incorrect is_tx_invalid flag.
+        // instruction.constrain_equal(is_tx_invalid, invalid_tx)
+        cb.condition(
+            balance_not_enough.expr() + (1.expr() - is_nonce_valid.expr()),
```

Comment:
why not just do
```rust
 cb.require_equal(
    "is_tx_invalid is correct",
    1 - (1 - balance_not_enough.expr()) * (is_nonce_valid.expr()),
    is_tx_invalid.expr())
```

Reply:
Can we use `*` in lhs?

Reply:
```rust
 cb.require_equal(
    "is_tx_invalid is correct",
    1.expr() - (1.expr() - balance_not_enough.expr()) * (is_nonce_valid.expr()),
    is_tx_invalid.expr())
```

returns an error
```
running 1 test
thread 'evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple' panicked at 'Too many bytes to compose an integer in field', zkevm-circuits/src/evm_circuit/util.rs:407:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/panicking.rs:142:14
   2: zkevm_circuits::evm_circuit::util::from_bytes::expr
             at ./src/evm_circuit/util.rs:407:9
   3: <zkevm_circuits::evm_circuit::execution::begin_tx::BeginTxGadget<F> as zkevm_circuits::evm_circuit::execution::ExecutionGadget<F>>::configure::{{closure}}
             at ./src/evm_circuit/execution/begin_tx.rs:163:17
   4: zkevm_circuits::evm_circuit::util::constraint_builder::ConstraintBuilder<F>::condition
             at ./src/evm_circuit/util/constraint_builder.rs:1258:19
   5: <zkevm_circuits::evm_circuit::execution::begin_tx::BeginTxGadget<F> as zkevm_circuits::evm_circuit::execution::ExecutionGadget<F>>::configure
             at ./src/evm_circuit/execution/begin_tx.rs:160:9
   6: zkevm_circuits::evm_circuit::execution::ExecutionConfig<F>::configure_gadget
             at ./src/evm_circuit/execution.rs:572:13
   7: zkevm_circuits::evm_circuit::execution::ExecutionConfig<F>::configure
             at ./src/evm_circuit/execution.rs:422:30
   8: zkevm_circuits::evm_circuit::EvmCircuit<F>::configure
             at ./src/evm_circuit.rs:48:34
   9: <zkevm_circuits::evm_circuit::test::TestCircuit<F> as halo2_proofs::plonk::circuit::Circuit<F>>::configure
             at ./src/evm_circuit.rs:261:31
  10: zkevm_circuits::evm_circuit::test::TestCircuit<F>::get_num_rows_required
             at ./src/evm_circuit.rs:331:26
  11: zkevm_circuits::evm_circuit::test::run_test_circuit
             at ./src/evm_circuit.rs:372:43
  12: zkevm_circuits::evm_circuit::test::run_test_circuit_geth_data_default
             at ./src/evm_circuit.rs:352:9
  13: zkevm_circuits::evm_circuit::execution::begin_tx::test::test_ok
             at ./src/evm_circuit/execution/begin_tx.rs:549:20
  14: zkevm_circuits::evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple
             at ./src/evm_circuit/execution/begin_tx.rs:569:9
  15: zkevm_circuits::evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple::{{closure}}
             at ./src/evm_circuit/execution/begin_tx.rs:567:5
  16: core::ops::function::FnOnce::call_once
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/ops/function.rs:248:5
  17: core::ops::function::FnOnce::call_once
             at /rustc/015a824f2dffe32707fceb59c47effaf7b73486c/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test evm_circuit::execution::begin_tx::test::begin_tx_gadget_simple ... FAILED
```

Reply:
can do, expression itself support such operation. but maybe not if considering the degree, need some analysis on this.